tun/tun_illumos.go

```diff
@@ -0,0 +1,519 @@
+	fmt.Printf("ip_muxid %v\n", ip_muxid)
+
+	return ip_muxid, nil
+}
+
+func unit_select(fd int, ppa int) (error) {
+	int_ppa := make([]byte, 4) /* storage for the PPA number */
+	*(*int32)(unsafe.Pointer(&int_ppa[0])) = int32(ppa)
+
+	_, err := ioctl(fd, IF_UNITSEL, uintptr(unsafe.Pointer(&int_ppa[0])))
+	if err != nil {
+		return fmt.Errorf("could not select unit: %v", err)
+	}
+
+	return nil
+}
+
+func push_ip(fd int) (error) {
+	/*
+	 * We need a C string with the value "ip".
+	 * XXX Is a Golang "string" implicitly NUL-terminated?
+	 */
+	modname := []byte{ 'i', 'p', 0 }
+
+	_, err := ioctl(fd, I_PUSH, uintptr(unsafe.Pointer(&modname[0])))
+	if err != nil {
+		return fmt.Errorf("could not push IP module: %v\n", err)
+	}
+
+	return nil
+}
+
+/*
+ * The "tun" device is a STREAMS module.  We need to make a STREAMS ioctl
+ * request to that module using the TUNNEWPPA command.  Returns the newly
+ * allocated PPA number.
+ */
+func tun_new_ppa(fd int) (int, error) {
+	for try_ppa := 0; try_ppa < 128; try_ppa++ {
+		/*
+		 * The data pointer (ic_dp) for a TUNNEWPPA request must point to an
+		 * int32_t value.  This value will be read to determine whether we want
+		 * to allocate a specific PPA number, or if we want a dynamically
+		 * assigned PPA number by passing -1.  If successful, the ioctl will
+		 * return the allocated PPA number.
+		 */
+		int_ppa := make([]byte, 4) /* storage for the PPA number */
+		*(*int32)(unsafe.Pointer(&int_ppa[0])) = int32(try_ppa)
+
+		/*
+		 * Construct a "struct strioctl" for use with the I_STR request
+		 * we will make to the "tun" device.
+		 */
+		strioc := make([]byte, 0x18) /* struct strioctl */
+		*(*int32)(unsafe.Pointer(&strioc[0])) = TUNNEWPPA /* int ic_cmd */
+		*(*int32)(unsafe.Pointer(&strioc[4])) = 0 /* int ic_timout */
+		*(*int32)(unsafe.Pointer(&strioc[8])) = 4 /* int ic_len */
+		*(*uintptr)(unsafe.Pointer(&strioc[16])) = /* int ic_dp */
+		    uintptr(unsafe.Pointer(&int_ppa[0]))
+
+		new_ppa, err := ioctl(fd, I_STR, uintptr(unsafe.Pointer(&strioc[0])))
+		if err == unix.EEXIST {
+			/*
+			 * This PPA appears to be in use; try the next one.
+			 */
+			continue
+		} else if err != nil {
+			return -1, fmt.Errorf("PPA allocation failure: %v", err)
+		}
+
+		return new_ppa, nil
+	}
+
+	return -1, fmt.Errorf("PPA allocation failure: all PPAs are busy")
+}
+
+/*
+ * Read bytes from the TUN device into this slice, and return the number of
+ * bytes we read.
+ */
+func (tun *NativeTun) read_tun(buf []byte) (int, error) {
+	// struct strbuf { /* 0x10 bytes */
+	//         int maxlen; /* offset: 0 bytes */
+	//         int len; /* offset: 4 bytes */
+	//         caddr_t buf; /* offset: 8 bytes */
+	// };
+	sbuf := make([]byte, 0x10) /* struct strbuf */
+	*(*int32)(unsafe.Pointer(&sbuf[0])) = int32(len(buf)) /* int maxlen */
+	*(*uintptr)(unsafe.Pointer(&sbuf[8])) = /* caddr_t buf */
+	    uintptr(unsafe.Pointer(&buf[0]))
+
+	var flags int32 = 0
+
+	_, err := getmsg(int(tun.tunFile.Fd()), uintptr(0),
```

Comment:
Do STREAMS offer any kind of batched read/write API? Or does getmsg/putmsg handle batched delivery under the hood? I ask because "one syscall per packet" ends up limiting tuntap performance on many platforms, and I'm wondering if illumos' use of STREAMS would let it sidestep that from the start.

Reply:
@gco pointed me at `bufmod(4M)`
- https://docs.oracle.com/cd/E88353_01/html/E37851/bufmod-4m.html
- https://illumos.org/man/4M/bufmod

I haven't explored what implementing that would look like here.

Reply:
Sounds like a nice optimization. Let's get things working right the vanilla way before we do things the wild way, though. It's already been a few years and this still hasn't shipped.

Reply:
Agreed. I just wanted to leave the breadcrumb for that here rather than leaving it sitting in my Twitter DMs. :)

Reply:
It will add latency, though.   I suspect that putting the STREAM in RMSGD mode and using readv with sufficiently large buffers will work to read multiple packets with one syscall.